File: artifacts/api-server/src/routes/ledger.ts
import { Router } from "express";
import { eq, and, sql } from "drizzle-orm";
import { db, inventoryLedgerTable, productsTable, warehousesTable } from "@workspace/db";
import { ListLedgerQueryParams } from "@workspace/api-zod";

const router = Router();

router.get("/ledger", async (req, res): Promise<void> => {
  const query = ListLedgerQueryParams.safeParse(req.query);
  if (!query.success) {
    res.status(400).json({ error: query.error.message });
    return;
  }
  const { productId, warehouseId, page = 1, limit = 50 } = query.data;
  const offset = (page - 1) * limit;

  const rows = await db
    .select({
      id: inventoryLedgerTable.id,
      productId: inventoryLedgerTable.productId,
      warehouseId: inventoryLedgerTable.warehouseId,
      movementId: inventoryLedgerTable.movementId,
      openingQty: inventoryLedgerTable.openingQty,
      changeQty: inventoryLedgerTable.changeQty,
      closingQty: inventoryLedgerTable.closingQty,
      transactionTime: inventoryLedgerTable.transactionTime,
      product: {
        id: productsTable.id,
        skuCode: productsTable.skuCode,
        productName: productsTable.productName,
        category: productsTable.category,
        brand: productsTable.brand,
        unit: productsTable.unit,
        minimumStock: productsTable.minimumStock,
        reorderStock: productsTable.reorderStock,
        isActive: productsTable.isActive,
        createdAt: productsTable.createdAt,
        updatedAt: productsTable.updatedAt,
      },
      warehouse: {
        id: warehousesTable.id,
        warehouseName: warehousesTable.warehouseName,
        location: warehousesTable.location,
        warehouseType: warehousesTable.warehouseType,
        createdAt: warehousesTable.createdAt,
      },
    })
    .from(inventoryLedgerTable)
    .innerJoin(productsTable, eq(inventoryLedgerTable.productId, productsTable.id))
    .innerJoin(warehousesTable, eq(inventoryLedgerTable.warehouseId, warehousesTable.id))
    .where(
      and(
        productId ? eq(inventoryLedgerTable.productId, productId) : undefined,
        warehouseId ? eq(inventoryLedgerTable.warehouseId, warehouseId) : undefined,
      )
    )
    .orderBy(sql`${inventoryLedgerTable.transactionTime} desc`)
    .limit(limit)
    .offset(offset);

  const countResult = await db
    .select({ count: sql<number>`count(*)::int` })
    .from(inventoryLedgerTable)
    .where(
      and(
        productId ? eq(inventoryLedgerTable.productId, productId) : undefined,
        warehouseId ? eq(inventoryLedgerTable.warehouseId, warehouseId) : undefined,
      )
    );

  res.json({ data: rows, total: countResult[0].count, page, limit });
});

export default router;

File: api/routers/ledger.py
from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session
from typing import Optional
from database import get_db
from models import InventoryLedger, Product, Warehouse
from schemas import LedgerOut

router = APIRouter(prefix="/api/ledger", tags=["ledger"])


@router.get("")
def list_ledger(
    product_id: Optional[int] = None,
    warehouse_id: Optional[int] = None,
    page: int = Query(1, ge=1),
    limit: int = Query(50, ge=1, le=200),
    db: Session = Depends(get_db),
):
    q = db.query(InventoryLedger).join(Product).join(Warehouse)
    if product_id:
        q = q.filter(InventoryLedger.product_id == product_id)
    if warehouse_id:
        q = q.filter(InventoryLedger.warehouse_id == warehouse_id)

    total = q.count()
    items = q.order_by(InventoryLedger.transaction_time.desc()).offset((page - 1) * limit).limit(limit).all()
    return {"data": [LedgerOut.model_validate(i) for i in items], "total": total, "page": page, "limit": limit}
